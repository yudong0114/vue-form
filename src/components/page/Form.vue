<template>
  <div class="form">
    <h1>Vue Form</h1>
    <InputItem
      itemName="名前"
      itemId="name"
      :itemValue="formData.name"
      @updateFormData="updateFormData"
    />
    <InputItem
      itemName="メールアドレス"
      itemId="email"
      :itemValue="formData.email"
      @updateFormData="updateFormData"
    />
  </div>
</template>
<style lang="scss" scoped>
.form {
  display: flex;
  flex-flow: column;
  align-items: center;
}
</style>
<script>
import InputItem from '@/components/organism/InputItem.vue';

export default {
  name: 'Form',
  components: {
    InputItem,
  },
  data() {
    return {
      formData: {
        name: '',
        email: '',
      },
    };
  },
  methods: {
    updateFormData(key, value) {
      this.formData[key] = value;
    },
  },
};
</script>
