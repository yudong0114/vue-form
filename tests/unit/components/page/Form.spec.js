import { shallowMount } from '@vue/test-utils';
import Form from '@/components/page/Form.vue';

describe('@/components/page/Form.vue', () => {
  it('スナップショットテスト', () => {
    const wrapper = shallowMount(Form, {
      propsData: {
        formData: {
          name: '',
          email: '',
        },
        formValidation: {
          name: [],
          email: [],
        },
        isSubmit: true,
        updateFormData: () => {},
        submit: () => {},
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it('Props設定時のコンポーネントの状態', () => {
    const wrapper = shallowMount(Form, {
      propsData: {
        formData: {
          name: '',
          email: '',
        },
        formValidation: {
          name: [],
          email: [],
        },
        isSubmit: true,
        updateFormData: () => {},
        submit: () => {},
      },
    });
    expect(wrapper.findAllComponents({ name: 'InputItem' })).toHaveLength(2);
    expect(wrapper.findComponent({ name: 'Button' })).toBeTruthy();
    expect(wrapper.findComponent({ name: 'Button' }).attributes().disabled).toBeFalsy();
  });
  it('バリデーションエラー時のテスト', () => {
    const wrapper = shallowMount(Form, {
      propsData: {
        formData: {
          name: '',
          email: '',
        },
        formValidation: {
          name: ['error'],
          email: ['error'],
        },
        isSubmit: false,
        updateFormData: () => {},
        submit: () => {},
      },
    });
    expect(wrapper.findAllComponents({ name: 'InputItem' })).toHaveLength(2);
    expect(wrapper.findAllComponents({ name: 'ErrorMessage' })).toHaveLength(2);
    expect(wrapper.findComponent({ name: 'Button' }).attributes().disabled).toBeTruthy();
  });
  it('ボタン状態のテスト', async () => {
    const wrapper = shallowMount(Form, {
      propsData: {
        formData: {
          name: '',
          email: '',
        },
        formValidation: {
          name: ['error'],
          email: ['error'],
        },
        isSubmit: true,
        updateFormData: () => {},
        submit: () => {},
      },
    });
    // isSubmitがtrueの状態 -> disabledが有効
    expect(wrapper.findComponent({ name: 'Button' }).attributes().disabled).toBeFalsy();
    // isSubmitがfalseの状態 -> disabledが無効
    wrapper.setProps({ isSubmit: false });
    await wrapper.vm.$nextTick();
    expect(wrapper.findComponent({ name: 'Button' }).attributes().disabled).toBeTruthy();
  });
  it('イベントによるメソッド発火のテスト', () => {
    const wrapper = shallowMount(Form, {
      propsData: {
        formData: {
          name: '',
          email: '',
        },
        formValidation: {
          name: [],
          email: [],
        },
        isSubmit: true,
        updateFormData: jest.fn(),
        submit: jest.fn(),
      },
    });
    // updateFormDataメソッド
    const spyUpdateFormData = jest.spyOn(wrapper.vm, 'updateFormData');
    // スタブコンポーネントからのeventUp
    wrapper.findComponent({ name: 'InputItem' }).vm.$emit('updateFormData');
    expect(spyUpdateFormData).toHaveBeenCalledTimes(1);
    // Propsのメソッドを実行
    wrapper.vm.updateFormData();
    expect(spyUpdateFormData).toHaveBeenCalledTimes(2);
    // submitメソッド
    const spySubmit = jest.spyOn(wrapper.vm, 'submit');
    // スタブコンポーネントからのeventUp
    wrapper.findComponent({ name: 'Button' }).vm.$emit('btnEmit');
    expect(spySubmit).toHaveBeenCalledTimes(1);
    // Propsのメソッドを実行
    wrapper.vm.submit();
    expect(spySubmit).toHaveBeenCalledTimes(2);
  });
});
