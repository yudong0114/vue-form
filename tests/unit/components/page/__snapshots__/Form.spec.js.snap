// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@/components/page/Form.vue スナップショットテスト 1`] = `
<div
  class="form"
>
  <h1>
    Vue Form
  </h1>
   
  <inputitem-stub
    itemid="name"
    itemname="名前"
    itemvalue=""
  />
    
  <inputitem-stub
    itemid="email"
    itemname="メールアドレス"
    itemvalue=""
  />
    
  <button-stub
    buttonlabel="フォームを送信"
  />
</div>
`;
